libsolidity/lsp/Utils.cpp

```diff
@@ -0,0 +1,70 @@
+#include <liblangutil/CharStreamProvider.h>
+#include <liblangutil/Exceptions.h>
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/FileRepository.h>
+#include <libsolidity/lsp/Utils.h>
+
+namespace solidity::lsp
+{
+
+using namespace frontend;
+using namespace langutil;
+using namespace std;
+
+optional<LineColumn> parseLineColumn(Json::Value const& _lineColumn)
+{
+	if (_lineColumn.isObject() && _lineColumn["line"].isInt() && _lineColumn["character"].isInt())
+		return LineColumn{_lineColumn["line"].asInt(), _lineColumn["character"].asInt()};
+	else
+		return nullopt;
+}
+
+Json::Value toJson(LineColumn _pos)
+{
+	Json::Value json = Json::objectValue;
+	json["line"] = max(_pos.line, 0);
+	json["character"] = max(_pos.column, 0);
+
+	return json;
+}
+
+Json::Value toJsonRange(LineColumn const& _start, LineColumn const& _end)
+{
+	Json::Value json;
+	json["start"] = toJson(_start);
+	json["end"] = toJson(_end);
+	return json;
+}
+
+vector<Declaration const*> allAnnotatedDeclarations(Expression const* _expression)
+{
+	vector<Declaration const*> output;
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_expression))
```

Comment:
Can you check how virtual lookup works in this case? Is it at least relative to the location of the expression or will it always go to the base?

Reply:
IIUC, I've been including a test case for that, which is exactly like clangd jumps. If you've meant something else, we can talk about ASAP then.